core/sr-primitives/src/lib.rs

```diff
@@ -44,6 +43,16 @@ pub mod traits;
 pub mod generic;
 pub mod transaction_validity;
 
+/// Full block error message.
+///
+/// This allows modules to indicate that given transaction is potentially valid
+/// in the future, but can't be executed in the current state.
+/// Note this error should be returned early in the execution to prevent DoS,
+/// cause the fees are not being paid if this error is returned.
+///
+/// Example: block gas limit is reached (the transaction can be retried in the next block though).
+pub const BLOCK_FULL: &str = "block size limit is reached";
```

Comment:
Wouldn't it be better to modify `ApplyOutcome` and add `BlockLimitReached` variant?

Reply:
I thought about this, but we need to return an `Err` from executive, not `Ok()` cause otherwise the changes won't be reverted. And we need to intercept the stringy error message from the module anyway.

Reply:
Call `map_api_result` in `block_builder.push_extrinsic` to map `BlockLimitReached` to an `Err` :)

Reply:
And then the result will be reverted :)

Reply:
It already calls `map_api_result` internally :) I think that was also the use case why I added `map_api_result`^^

Reply:
After looking into the function, why is `ApplyOutcome::Fail` not interpreted as error?

Reply:
Cause that's the exact distinction between `Ok` and `Err`. 
In case of `Err`: we revert everything that happened in the current context (includes paying fees) and the extrinsic is not included in the block but rather marked as invalid. Errors should be returned early in the process to avoid DoS vectors. (invalid transactions are removed from the pool, cause the assumption is they want be part of the block anyway).
In case of `Ok`: the extrinsic still pays the fee, but it fails being dispatched because of some other reasons - this doesn't revert and the extrinsic is still part of the block.

IMHO block full is more a pre-condition of a module that failed and is an exceptional case hence we should return it as `Err` not `ApplyOutcome`.